### PR TITLE
feature: add architecture for security plugins

### PR DESCRIPTION
**Pull request by sbender9**
Adding a way for developers to add authentication/authorization to node server via a special security plugin

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -49,6 +49,12 @@ function Server(opts) {
   app.version = "0.0.1"
 
   app.use(require('body-parser').json())
+  
+  if ( app.config.settings.authentication && app.config.settings.authentication.plugin )
+  {
+    app.securityPlugin = require(app.config.settings.authentication.plugin)(app)
```
Comment (tkurki) [inline lib/index.js:55]: Why not use security in naming in both settings & app property?

File: lib/interfaces/ws.js [hunk 2]
```diff
@@ -43,11 +43,22 @@ module.exports = function(app) {
...
+    if ( app.securityPlugin )
+    {
+      app.securityPlugin.setupWS(primus)
+    }
+    
     primus.on('connection', function(spark) {
       debug(spark.id + " connected with params " + JSON.stringify(spark.query));
 
       var onChange = function(data) {
+        if ( app.securityPlugin && app.securityPlugin.verifyWS(spark) == false )
```
Comment (tkurki) [inline lib/interfaces/ws.js:56]: !app.securityPlugin.verifyWS(spark) would be more idiomatic

File: lib/interfaces/ws.js [hunk 3]
```diff
@@ -43,11 +43,22 @@ module.exports = function(app) {
...
+        {
+          spark.end(undefined, { reconnect: true });
```
Comment (tkurki) [inline lib/interfaces/ws.js:58]: Does the reconnect actually work if the client is Primus? Have you checked what it actually does?

How about a human oriented message down the connection before disconnecting? `{message: "Connection disconnected by security constraint"}` or something.

File: lib/interfaces/ws.js [hunk 4]
```diff
@@ -56,6 +67,12 @@ module.exports = function(app) {
       spark.on('data', function(msg) {
         debug("<" + JSON.stringify(msg));
         if (msg.updates) {
+          if ( app.securityPlugin && app.securityPlugin.shouldAllowWrite(spark) == false )
+
+          {
+            debug("security dissalowed update")
```
Comment (tkurki) [inline lib/interfaces/ws.js:73]: Typo

**Comment by tkurki** [comment]
More about naming: how about calling things `securityStrategy`?

Sorry for letting this sit for so long. I've been pushing this off for ever, but the way you have structured it turns out very uninvasive to the core. I assume you have a concrete implementation for this - I'd like to check that out. 

I think this needs a bit of documentation about the interface that the plugin must implement.

The flip side of uninvasiveness is that this is very unopinionated, whereas I should think we should have a security model that includes the different operations. But I am happy to merge this and then continue from there.

File: lib/interfaces/ws.js (hunk 3 above)
Comment (sbender9) [reply lib/interfaces/ws.js:58]: I've only tested with wscat and WilhelmSK, it doesn't do anything there. I'll test with Primus...

**Comment by sbender9** [comment]
I like the `securityStrategy`,does not cause confusion with the other plugins

https://github.com/sbender9/signalk-jwt-security

https://github.com/sbender9/signalk-jwt-security-config

**Comment by sbender9** [comment]
Replaced by #248

**Pull request closed without merging**